### PR TITLE
Redesign Circle Ci pipelines

### PR DESCRIPTION
**Pull request by rjonczy**
## Purpose

- separate helm builds and core taraxa-node builds
- separate release builds from development builds

## How does the solution address the problem

**Workflows**:
- `build-docker-image`:
   - _description_: builds **development docker image** of`taraxa-node`, test and publish to **GCR**
   - _triggers_: every commit in branch, apart from branches related to helm charts (`chart/`, `chore/chart-`, `fix/chart-`, `feature/chart-`)
- `release-docker-image`:
   - _description_: 
   - _triggers_: tags:
     - `vX.Y.Z`
     - `vX.Y.Z-beta` 
- `build-helm-chart`:
   - _description_: builds and publish **development version of helm chart** (publish versions `X.Y.Z-sha.<short_commit_sha>` taken from Chart.yaml in combination with short commit sha, i.e. `0.1.1-sha.9a93ea9`)
   - _triggers_: every commit in branch with PR starting with:
     - `chart/`
     - `chore/chart-`
     - `fix/chart-`
     - `feature/chart-`
- `release-helm-chart`:
   - _description_: builds and publish **release version of helm chart** (releases versions of chart `X.Y.Z` taken from tag, must be equal to version in `Chart.yaml` i.e. `0.1.1`)
   - _triggers_: tags starting with:
     - `chart-v` i.e. `chart-v1.2.3`
- `redeploy-prnet`: not modified
- `cleanup-prnet`: not modified
- `create-builder-image`: not modified

**Artifacts from branches**: 

- Docker Images

| Name  | Artifact  | Repository |
|---|---|---|
| `develop`  | `gcr.io/jovial-meridian-249123/taraxa-node-develop:${CIRCLE_SHA1}`  |  GCR |
| `develop`  | `gcr.io/jovial-meridian-249123/taraxa-node-develop:${SHORT_GIT_HASH}` |  GCR |
| `develop`  |  `gcr.io/jovial-meridian-249123/taraxa-node-develop:${DOCKER_BRANCH_TAG}-${CIRCLE_BUILD_NUM}` |  GCR |
| `develop`  |  `gcr.io/jovial-meridian-249123/taraxa-node-develop:${START_TIME}` |  GCR |
| `develop`  |  `gcr.io/jovial-meridian-249123/taraxa-node-develop:${START_TIME}-${CIRCLE_SHA1}` |  GCR |
| `develop`  |  `gcr.io/jovial-meridian-249123/taraxa-node-develop:latest` |  GCR |
| `release/vX.Y.Z`  | `gcr.io/jovial-meridian-249123/taraxa-node-release:${CIRCLE_SHA1}`  |  GCR |
| `release/vX.Y.Z`  | `gcr.io/jovial-meridian-249123/taraxa-node-release:${SHORT_GIT_HASH}` |  GCR |
| `release/vX.Y.Z`  |  `gcr.io/jovial-meridian-249123/taraxa-node-release:${DOCKER_BRANCH_TAG}-${CIRCLE_BUILD_NUM}` |  GCR |
| `release/vX.Y.Z`  |  `gcr.io/jovial-meridian-249123/taraxa-node-release:${START_TIME}` |  GCR |
| `release/vX.Y.Z`  |  `gcr.io/jovial-meridian-249123/taraxa-node-release:${START_TIME}-${CIRCLE_SHA1}` |  GCR |
| `release/vX.Y.Z`  |  `gcr.io/jovial-meridian-249123/taraxa-node-release:latest` |  GCR |
| `*` + PR  |  `gcr.io/jovial-meridian-249123/taraxa-node:pr-${PR}-${CIRCLE_BUILD_NUM}` |  GCR |
| `*` + PR  |  `gcr.io/jovial-meridian-249123/taraxa-node:pr-${PR}-${CIRCLE_SHA1}` |  GCR |
| `*` + PR  |  `gcr.io/jovial-meridian-249123/taraxa-node:pr-${PR}-${SHORT_GIT_HASH}` |  GCR |

- Helm Charts

| Name  | Artifact  | Repository |
|---|---|---|
| `/^chart\/.*/` | `X.Y.Z-sha.${SHORT_GIT_HASH}` |  charts.gcp.taraxa.io |
| `/^chore\/chart-.*/` | `X.Y.Z-sha.${SHORT_GIT_HASH}` |  charts.gcp.taraxa.io |
| `/^fix\/chart-.*/` | `X.Y.Z-sha.${SHORT_GIT_HASH}` |  charts.gcp.taraxa.io |
| `/^feature\/chart-.*/` | `X.Y.Z-sha.${SHORT_GIT_HASH}` |  charts.gcp.taraxa.io |

**Artifacts from tags**: 
- Docker Images

| Name  | Pattern  | Repository |
|---|---|---|
| `vX.Y.Z`  |  `taraxa/taraxa-node:${CIRCLE_TAG}` | Docker Hub |
| `vX.Y.Z-beta`  |  `taraxa/taraxa-node:${CIRCLE_TAG}` | Docker Hub |

- Helm Charts

| Name  | Pattern  | Repository |
|---|---|---|
| `chart-vX.Y.Z`  |  `taraxa/taraxa-node:${CIRCLE_TAG}` | charts.gcp.taraxa.io |

## Changes made

- Changes to Circle Ci pipelines (`.circleci/config.yml`)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -426,7 +427,111 @@ jobs:
           comment: "Failed to Clean up prnet"
           pr_number: << pipeline.parameters.pr_number >>
 
-  build-and-push-docker-image:
+  build-helm-chart:
+    environment:
+      - IMAGE: taraxa-node
+      - HELMCHART: taraxa-node
+    machine:
+      image: ubuntu-2204:2022.04.2
```
Comment (rattrap) [inline .circleci/config.yml:435]: No need to use this image anymore. This was used for the node build I guess but now we can just use the helm image, right?

File: .circleci/config.yml [hunk 2]
```diff
@@ -426,7 +427,111 @@ jobs:
...
+      docker_layer_caching: true
+    resource_class: medium
+    steps:
+      - checkout
+
+      - run:
+          name: Prepare Environment
+          command: |
+            echo "export PR=$( [[ -z ${CIRCLE_PULL_REQUEST+x} ]] && echo false || echo $(basename ${CIRCLE_PULL_REQUEST}) )" >> $BASH_ENV
```
Comment (rattrap) [inline .circleci/config.yml:444]: Don't really like this repeating part where we set env vars. Maybe we can create a yaml ref for this.

File: .circleci/config.yml [hunk 3]
```diff
@@ -426,7 +427,111 @@ jobs:
...
+            export PR=$( [[ -z ${CIRCLE_PULL_REQUEST+x} ]] && echo false || echo $(basename ${CIRCLE_PULL_REQUEST}) )
+            echo "export HELM_TEST_NAME=pr-${PR}" >> $BASH_ENV
+            curl https://baltocdn.com/helm/signing.asc | sudo apt-key add -
+            sudo apt-get install apt-transport-https --yes
+            echo "deb https://baltocdn.com/helm/stable/debian/ all main" | sudo tee /etc/apt/sources.list.d/helm-stable-debian.list
+            sudo apt-get update
+            sudo apt-get install helm
+            helm plugin install https://github.com/chartmuseum/helm-push.git
+            helm repo add --username ${HELM_REPO_USERNAME} --password ${HELM_REPO_PASSWORD} taraxacharts https://charts.gcp.taraxa.io
+            helm repo add bitnami https://charts.bitnami.com/bitnami
```
Comment (rattrap) [inline .circleci/config.yml:455]: Why do we need bitnami here?

File: .circleci/config.yml [hunk 4]
```diff
@@ -426,7 +427,111 @@ jobs:
...
+            sudo cp kubectl /usr/local/bin 
+            sudo chmod +x /usr/local/bin/kubectl
+            sudo wget -qO /usr/local/bin/yq https://github.com/mikefarah/yq/releases/latest/download/yq_linux_amd64
+            sudo chmod a+x /usr/local/bin/yq
+
+      - run:
+          name: Verify Helm chart with linter
+          command: |
+            if [[ ${PR} != "false" ]]; then
+              helm lint charts/${HELMCHART}/
```
Comment (rattrap) [inline .circleci/config.yml:469]: Why not use the new orb lint functionality?

File: .circleci/config.yml [hunk 5]
```diff
@@ -426,7 +427,111 @@ jobs:
...
+                --timeout 600s \
+                --namespace ${HELM_TEST_NAME:0:40}
+            fi
+
+      - run:
+          name: Publish chart to Taraxa's repository
+          command: |
+            HELMCHART_VERSION=$(yq .version charts/${HELMCHART}/Chart.yaml | tr -d \")
+            if [[ ${PR} != "false" ]]; then
+              helm cm-push charts/${HELMCHART}/ --version="${HELMCHART_VERSION}-sha.${SHORT_GIT_HASH}" taraxacharts || true
```
Comment (rattrap) [inline .circleci/config.yml:522]: Hm. Wasn't the purpose using the new orb here?

File: .circleci/config.yml [hunk 6]
```diff
@@ -426,7 +427,111 @@ jobs:
...
+          command: |
+            if [[ ${PR} != "false" ]]; then
+              helm -n ${HELM_TEST_NAME:0:40} delete ${HELM_TEST_NAME:0:40} || true
+              kubectl delete ns ${HELM_TEST_NAME:0:40} || true
+            fi
+
+  build-docker-image:
     environment:
       - IMAGE: taraxa-node
       - HELMCHART: taraxa-node
```
Comment (rattrap) [inline .circleci/config.yml:537]: Don't need this env anymore.

File: .circleci/config.yml [hunk 7]
```diff
@@ -457,12 +562,12 @@ jobs:
...
+            echo "export SHORT_GIT_HASH=$(echo $CIRCLE_SHA1 | cut -c -7)" >> $BASH_ENV
             echo "export VERSION=$(git rev-parse --short HEAD)" >> $BASH_ENV
             echo "export START_TIME=$(date +%Y%m%d-%Hh%Mm%Ss)" >>$BASH_ENV
             echo "export GCP_IMAGE=gcr.io/${GOOGLE_PROJECT_ID}/${IMAGE}" >> $BASH_ENV
             echo "export PR=$( [[ -z ${CIRCLE_PULL_REQUEST+x} ]] && echo false || echo $(basename ${CIRCLE_PULL_REQUEST}) )" >>$BASH_ENV
             export PR=$( [[ -z ${CIRCLE_PULL_REQUEST+x} ]] && echo false || echo $(basename ${CIRCLE_PULL_REQUEST}) )
-            echo "export HELM_TEST_NAME=pr-${PR}" >> $BASH_ENV
             sudo service apport stop
             curl https://baltocdn.com/helm/signing.asc | sudo apt-key add -
             sudo apt-get install apt-transport-https --yes
```
Comment (rattrap) [inline .circleci/config.yml:573]: Don't need these tools anymore. These are just for the helm part.

File: .circleci/config.yml [hunk 8]
```diff
@@ -649,94 +755,187 @@ jobs:
...
 
-      - run:
-          name: Generate Helm chart and output yaml manifests for verification
-          command: |
-            helm dependency build charts/${HELMCHART}/
-            helm template charts/${HELMCHART}/ --values charts/${HELMCHART}/values.yaml
+  release-docker-image:
+      environment:
+        - IMAGE: taraxa-node
+        - HELMCHART: taraxa-node
```
Comment (rattrap) [inline .circleci/config.yml:772]: Don't need this env var anymore.

File: .circleci/config.yml [hunk 9]
```diff
@@ -649,94 +755,187 @@ jobs:
...
+              helm repo add --username ${HELM_REPO_USERNAME} --password ${HELM_REPO_PASSWORD} taraxacharts https://charts.gcp.taraxa.io
+              helm repo add bitnami https://charts.bitnami.com/bitnami
+              helm repo update
+              mkdir -p $HOME/.kube
+              echo -n "${KUBE_CONFIG}" | base64 --decode > $HOME/.kube/config
+              curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl"
+              sudo cp kubectl /usr/local/bin 
+              sudo chmod +x /usr/local/bin/kubectl
+              sudo wget -qO /usr/local/bin/yq https://github.com/mikefarah/yq/releases/latest/download/yq_linux_amd64
+              sudo chmod a+x /usr/local/bin/yq
```
Comment (rattrap) [inline .circleci/config.yml:821]: Don't need this anymore.

**Review by rattrap — CHANGES_REQUESTED** [review]

File: .circleci/config.yml (hunk 2 above)
Comment (rjonczy) [reply .circleci/config.yml]: Moved to separate `command` of Circle Ci, in which are :
- defined env variable for later tasks
- installed tools (`jq`, `yq`, ...)
- cleaning old images, ...

File: .circleci/config.yml [hunk 10]
```diff
@@ -426,7 +427,111 @@ jobs:
           comment: "Failed to Clean up prnet"
           pr_number: << pipeline.parameters.pr_number >>
 
-  build-and-push-docker-image:
+  build-helm-chart:
+    environment:
+      - IMAGE: taraxa-node
```
[the remark this answers is not in the document]
Comment (rjonczy) [reply .circleci/config.yml]: Refactored. This job was removed.

File: .circleci/config.yml (hunk 1 above)
Comment (rjonczy) [reply .circleci/config.yml]: Refactored. This job was removed.

2 workflows realasing helm charts:
- development
- production 
released operate on the same private orb.

```
  # run this workflow for branches specified below
  build-helm-chart:
    jobs:
      - helm-release/chart-publish:
          develop: true # means, that this is development release, no-verify and artifact -> ${HELMCHART_VERSION}-sha.${SHORT_GIT_HASH}
          charts-dir: charts
          chart-name: taraxa-node
          chartmuseum-url: https://charts.gcp.taraxa.io
          chartmuseum-username: HELM_REPO_USERNAME
          chartmuseum-password: HELM_REPO_PASSWORD
          context:
            - CHARTSREPO
          filters:
            branches:
              only: 
              - /^chart\/.*/
              - /^chore\/chart-.*/
              - /^fix\/chart-.*/
              - /^feature\/chart-.*/


  # run this workflow for tags, like chart-vX.Y.Z
  release-helm-chart:
    jobs:
      - helm-release/chart-publish:
          charts-dir: charts
          chart-name: taraxa-node
          chartmuseum-url: https://charts.gcp.taraxa.io
          chartmuseum-username: HELM_REPO_USERNAME
          chartmuseum-password: HELM_REPO_PASSWORD
          context:
            - CHARTSREPO
          filters:
            branches:
              ignore: /.*/
            tags:
              only: /^chart-v\d+.\d+.\d+/
```

File: .circleci/config.yml (hunk 3 above)
Comment (rjonczy) [reply .circleci/config.yml]: Refactored. 

`bitnami removed`

File: .circleci/config.yml (hunk 4 above)
Comment (rjonczy) [reply .circleci/config.yml]: Refactored. Using now orb. Look comment: https://github.com/Taraxa-project/taraxa-node/pull/2110/files/b7f7a7d322e7a8db10f37ba638067dab897ebd00#r992369570

File: .circleci/config.yml (hunk 6 above)
Comment (rjonczy) [reply .circleci/config.yml]: removed `HELMCHART` from this job

File: .circleci/config.yml (hunk 7 above)
Comment (rjonczy) [reply .circleci/config.yml]: Refactored.

Tools Installation was removed.

File: .circleci/config.yml (hunk 8 above)
Comment (rjonczy) [reply .circleci/config.yml]: Removed.

File: .circleci/config.yml (hunk 9 above)
Comment (rjonczy) [reply .circleci/config.yml]: Refactored. Removed

File: .circleci/config.yml [hunk 11]
```diff
@@ -297,6 +298,244 @@ commands:
...
+    steps:
+      - run:
+          name: Login into Docker Hub
+          command: |
+            echo ${DOCKERHUB_PASS} | docker login -u taraxa --password-stdin
+
+      - run:
+          name: Push Images
+          command: |
+            docker push taraxa/${IMAGE}:${CIRCLE_TAG}
```
Comment (rattrap) [inline .circleci/config.yml:459]: I think we should add the same conditions here. Only push if:
`if [[ ${CIRCLE_BRANCH} == "testnet" ]];then`
or
`if [[ ${CIRCLE_TAG} != "" ]];then`

**Review by rattrap — APPROVED** [review]

File: .circleci/config.yml [hunk 12]
```diff
@@ -1030,9 +975,83 @@ workflows:
...
-              only: /^v?\d+\.\d+\.\d+(?:-.*)?$/
+              ignore: /.*/     # so far not working, so do not run (revisit at some point)
           requires:
             - build-linux
             - build-mac
+
+  # run this workflow for all branches apart those reserved for chart
+  build-docker-image:
+    jobs:
+      - build-docker-image:
```
Comment (rattrap) [inline .circleci/config.yml:988]: So if the "testnet" branch is built using this action, we're not pushing to docker hub

**Review by rattrap — APPROVED** [review]

File: .circleci/config.yml (hunk 11 above)
Comment (rattrap) [reply .circleci/config.yml:459]: Nevermind. My bad.

File: .circleci/config.yml (hunk 12 above)
Comment (rattrap) [reply .circleci/config.yml:988]: Nevermind. My bad.

**Review by rattrap — APPROVED** [review]

**Review by rattrap — APPROVED** [review]

**Pull request merged**